dbms/src/Interpreters/MergeJoin.cpp

```diff
@@ -546,4 +725,30 @@ void MergeJoin::addRightColumns(Block & block, MutableColumns && right_columns)
     }
 }
 
+template <bool in_memory>
+size_t MergeJoin::rightBlocksCount()
+{
+    if constexpr (!in_memory)
+        return flushed_right_blocks.size();
+    else
+        return loaded_right_blocks.size();
+}
+
+template <bool in_memory>
```

Comment:
Why is it a template parameter and not just a normal parameter? At a first glance, it looks like this function works on the level of entire blocks, so we can just branch in runtime and don't have to statically generate two functions, and it won't lead to any performance problems.

Reply:
It's some technics to use template for passthrought strategy without additional cost. If I remember correctly it's "Template Method" without interfaces.
```
tempalte <int strategy>
dificult_func(...)
{
    x = simple_func1<strategy>()
    ....
    simple_func2<strategy>()
}
```

Reply:
Yep, that's how I read it. But it's not without cost -- it increases code size and compile times. Personally, I also find it more cumbersome to read. This code doesn't have strong performance requirements, and the strategy is a simple bool, so maybe it doesn't have to pass it statically, and passing dinamically will suffice?

Reply:
> This code doesn't have strong performance requirements
Not exactly. On-disk variant really is not perf critical here. The idea is to have zero cost for in-memory variant. It should be as fast as there's not on-disk variant at all. It has very simple loadRightBlock() that expected to be slower with additional if.

Reply:
Well, OK then.